aeneas/src/ir/PackingSolver.v3

```diff
@@ -0,0 +1,339 @@
+
+Based on each class, the packing solver is able to squeeze other fields into the
+unassigned portions of the representation.
+*/
+
+type Bit {
+	case Unassigned; // ?
+	case Fixed(v: u1); // 0 or 1
+	case Unusable; // .
+}
+
+ // TODO: Allow non-contiguous intervals to be represented
+type Interval(s: int, e: int) {
+	def render(sb: StringBuilder) -> StringBuilder {
+		return sb.put2("%d-%d", s, e);
+	}
+}
+
+component PackingPatterns {
+	def fromString(s: string) -> PackingPattern {
+		var bits = Vector<Bit>.new();
+		for (i < s.length) {
+			var c = s[s.length - i - 1];
+			match (c) {
+				_ => ;
+				'0' => bits.put(Bit.Fixed(0));
+				'1' => bits.put(Bit.Fixed(1));
+				'.' => bits.put(Bit.Unusable);
+				'?' => bits.put(Bit.Unassigned);
+			}
+		}
+		return PackingPattern.new(bits.extract());
+	}
+}
+
+class PackingPattern(bits: Array<Bit>) {
+	def size = bits.length;
+	
+	def render(s: StringBuilder) -> StringBuilder {
+		for (i < bits.length) {
+			var b = bits[bits.length - i - 1];
+			if (i != 0 && i % 8 == 0) s.putc('_');
+			match (b) {
+				Unassigned => s.putc('?');
+				Fixed(v) => s.put1("%d", v);
+				Unusable => {
+					// s.puts(Terminal.CTRL_GREEN);
+					// s.pututf8(0x25ae);
+					// s.puts(Terminal.CTRL_DEFAULT);
+					s.putc('.');
+				}
+			}
+		}
+		return s;
+	}
+	def getIntervalsForSize(s: int) -> Array<Interval> {
+		var i = 0;
+		var intervals = Vector<Interval>.new();
+
+		for (i < size - s + 1) {
+			var start = size - s - i;
+			var isEmpty = true;
+			for (j < s) {
+				if (bits[start + j] != Bit.Unassigned) {
+					isEmpty = false;
+					break;
+				}
+			}
+			if (isEmpty) intervals.put(Interval(start, start + s));
+		}
+		return intervals.extract();
+	}
+	def assignInterval(i: Interval) {
+		for (j = i.s; j < i.e; j++) bits[j] = Bit.Unusable;
+	}
+	def tryAssignInterval(i: Interval) -> bool {
+		for (j = i.s; j < i.e; j++) if (bits[j] != Bit.Unassigned) return false;
+		assignInterval(i);
+		return true;
+	}
+	def unassignInterval(i: Interval) {
+		for (j = i.s; j < i.e; j++) bits[j] = Bit.Unassigned;
+	}
+	def copy() -> PackingPattern { return PackingPattern.new(Arrays.dup(bits)); }
+}
+
+type PackingField {
+	case Nonref(size: byte);
+	case Ref;
+}
+
+type CaseField(caseIdx: int, fieldIdx: int) {
+	def hash() -> int { return caseIdx * 31 + fieldIdx; }
+}
+
+class PackingSolver(size: byte, refPatterns: RefPatterns) {
+	var cases: Array<Array<PackingField>>;
+	var fieldOrder: Array<CaseField>;
+	var assignments = HashMap<CaseField, Interval>.new(CaseField.hash, CaseField.==);
+
+	def reset() {
+		cases = null;
+		fieldOrder = null;
+		assignments = HashMap<CaseField, Interval>.new(CaseField.hash, CaseField.==);
+	}
+
+	def canDistinguish(state: Array<PackingPattern>, elements: Array<bool>) -> bool {
+		// TODO: split on bits with the most unassigned
+		var numElements = 0;
+		for (i in elements) if (i) numElements++;
+		if (numElements <= 1) return true;
+
+		for (i < size) {
```

Comment:
I think this will be more efficient if you do a first pass over all the bits and try to find a bit that is fixed in all patterns, in order to avoid duplicating an element when trying an assignment, ranking them otherwise along the way. If you find a bit fixed in all elements, stop and use it right away. If no bit fixed in all patterns is found, then use the best-ranked bit instead. The rank is some function of how many patterns have that unassigned versus fixed 0 or 1. E.g. picking a bit that is fixed in most patterns (but all 0) isn't particularly helpful;there's probably an information-theoretic best choice. But obviously it's a heuristic, so any choice is going to be bad in some situations.